dll/cpl/sysdm/environment.c

```diff
@@ -17,6 +25,114 @@ typedef struct _VARIABLE_DATA
     LPTSTR lpCookedValue;
 } VARIABLE_DATA, *PVARIABLE_DATA;
 
+typedef struct _DIALOG_DATA
+{
+    BOOL bIsFriendlyUI;
+    BOOL bIsItemSelected;
+    DWORD dwSelectedValueIndex;
+    DWORD cxMin;
+    DWORD cyMin;
+    DWORD cxOld;
+    DWORD cyOld;
+    HWND hEditBox;
+    PVARIABLE_DATA VarData;
+} DIALOG_DATA, *PDIALOG_DATA;
+
+static DWORD
+GatherDataFromEditBox(HWND hwndDlg,
+                      PVARIABLE_DATA VarData)
+{
+    DWORD dwNameLength;
+    DWORD dwValueLength;
+
+    dwNameLength = SendDlgItemMessage(hwndDlg, IDC_VARIABLE_NAME, WM_GETTEXTLENGTH, 0, 0);
+    dwValueLength = SendDlgItemMessage(hwndDlg, IDC_VARIABLE_VALUE, WM_GETTEXTLENGTH, 0, 0);
+
+    if (dwNameLength == 0 || dwValueLength == 0)
+    {
+        return 0;
+    }
+    if (VarData->lpName == NULL)
+    {
+        VarData->lpName = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
+    }
+    else if (_tcslen(VarData->lpName) < dwNameLength)
+    {
+        GlobalFree(VarData->lpName);
+        VarData->lpName = GlobalAlloc(GPTR, (dwNameLength + 1) * sizeof(TCHAR));
+    }
+    SendDlgItemMessage(hwndDlg, IDC_VARIABLE_NAME, WM_GETTEXT, dwNameLength + 1, (LPARAM)VarData->lpName);
+
+    if (VarData->lpRawValue == NULL)
+    {
+        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+    }
+    else if (_tcslen(VarData->lpRawValue) < dwValueLength)
+    {
+        GlobalFree(VarData->lpRawValue);
+        VarData->lpRawValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+    }
+    SendDlgItemMessage(hwndDlg, IDC_VARIABLE_VALUE, WM_GETTEXT, dwValueLength + 1, (LPARAM)VarData->lpRawValue);
+
+    return dwValueLength;
+}
+
+static DWORD
+GatherDataFromListView(HWND hwndListView,
+                      PVARIABLE_DATA VarData)
+{
+    DWORD dwValueLength;
+    DWORD NumberOfItems;
+    DWORD i;
+    TCHAR szData[MAX_PATH];
```

Comment:
That looks too small.

Reply:
@HBelusca Since that dialog box is meant for variables having multiple directories as "values", I decided to go for `MAX_PATH` value. (and what that is really for to gather item strings one by one)

Reply:
So to be clear, you can "safely" store only **one** path in this variable, right?
(btw. "safely" because it wouldn't guarantee to be able to hold a path of more than `MAX_PATH`)

Reply:
So what can be the maximum length that a directory can have. @Doug-Lyons @HBelusca 
Also yes, only one item element is stored in it at a time.

Reply:
Just use a dynamic-size buffer.

Reply:
> Just use a dynamic-size buffer.

However there is no direct way to gather the length of an item string. Hence why I went with this.
(Basically first gather the length of the strings one by one and then using it to allocate the `lpRawValue` and then adding the strings to it.)

Reply:
Then in order to be 100% sure there won't ever be corruption of the PATH directory list (due to potential truncation of paths to `MAX_PATH`, if a longer path is present in this environment variable), you need to do more active checks:
- When deciding whether or not to open the "advanced" (list-view) dialog for the PATH env, check somewhere whether there are paths in this environment variable that are longer than `MAX_PATH`. If so, fall back to the "old" dialog (using the textbox).
- If the "advanced" dialog is used, check whether the user is trying to add a path longer than `MAX_PATH`; if so, display an error messagebox telling he cannot add such long path.

Reply:
It's nice, this is already implementing my second point. And ok then subtract 1 from the limit-text length (as it doesn't count the null-terminator), and add a comment about that.

Reply:
> It's nice, this is already implementing my second point. And ok then subtract 1 from the limit-text length (as it doesn't count the null-terminator), and add a comment about that.

Now for your first point, I made sure that the "fancy" dialog box will open only when the variable has valid directories.